Sources/web3swift/Transaction/EIP1559Envelope.swift

```diff
@@ -0,0 +1,320 @@
+    /// - `Block.nextBlockBaseFeePerGas` - baseFee which will be burnt during the transaction processing
+    /// - `self.maxPriorityFeePerGas` - explicit amount of a tip to the miner of the given block which will include this transaction
+    ///
+    /// If amount of this will be **greater** than sum of `Block.baseFeePerGas` and `maxPriorityFeePerGas`
+    /// all exceed funds will be returned to the sender.
+    ///
+    /// If amount of this will be **lower** than sum of `Block.baseFeePerGas` and `maxPriorityFeePerGas`
+    /// miner will recieve amount calculated by the following equation: `maxFeePerGas - Block.baseFeePerGas`
+    /// where 'Block' is the block that the transaction will be included.
+    public var maxFeePerGas: BigUInt
+    public var accessList: [AccessListEntry] // from EIP-2930
+
+    private var internalChainID: BigUInt
+
+    // for CustomStringConvertible
+    public var description: String {
+        var toReturn = ""
+        toReturn += "Type: " + String(describing: self.type) + "\n"
+        toReturn += "chainID: " + String(describing: self.chainID) + "\n"
+        toReturn += "Nonce: " + String(describing: self.nonce) + "\n"
+        toReturn += "Gas limit: " + String(describing: self.gasLimit) + "\n"
+        toReturn += "Max priority fee per gas: " + String(describing: self.maxPriorityFeePerGas) + "\n"
+        toReturn += "Max fee per gas: " + String(describing: maxFeePerGas) + "\n"
+        toReturn += "To: " + self.to.address + "\n"
+        toReturn += "Value: " + String(describing: self.value) + "\n"
+        toReturn += "Data: " + self.data.toHexString().addHexPrefix().lowercased() + "\n"
+        toReturn += "Access List: " + String(describing: accessList) + "\n"
+        toReturn += "v: " + String(self.v) + "\n"
+        toReturn += "r: " + String(self.r) + "\n"
+        toReturn += "s: " + String(self.s) + "\n"
+        return toReturn
+    }
+}
+
+extension EIP1559Envelope {
+    private enum CodingKeys: String, CodingKey {
+        case chainId
+        case nonce
+        case to
+        case value
+        case maxPriorityFeePerGas
+        case maxFeePerGas
+        case gasLimit
+        case gas
+        case data
+        case input
+        case accessList
+        case v
+        case r
+        case s
+    }
+
+    public init?(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        if !container.contains(.to) || !container.contains(.nonce) || !container.contains(.value) || !container.contains(.chainId) { return nil }
+        if !container.contains(.data) && !container.contains(.input) { return nil }
+        if !container.contains(.v) || !container.contains(.r) || !container.contains(.s) { return nil }
+
+        // everything we need is present, so we should only have to throw from here
+        self.internalChainID = try container.decodeHexIfPresent(to: BigUInt.self, key: .chainId) ?? 0
```

Comment:
except this init is failable and throws if a decoding error happens. In this case since I have already tested for the presence of required fields, I could simply do a `try` and have it throw on the error, this would yield the same effect in this case.

The way you suggest, it would guarantee that the init succeeds as long as all fields are present, but even if the data associated with those fields is invalid.

Reply:
To be clear, my intent here is to start laying a path to where errors can be bubbled up to the app level to be handled in a meaningful way.